### DISCUSSION TITLE
How to extract original videos from the dataset?

### DISCUSSION THREAD

**Discussion by tctco** [talmolab/sleap#779] [Help!]
Hi, thanks for this exciting project!

Is there a way to get the original video from the dataset (namely the `slp` files)? I would like to compare the performance between SLEAP and my own algorithm.

Thanks in advance!

**Comment by roomrys** [comment]
Hi @tctco,

Super cool that you are developing your own algorithm! 

If you imported a `mp4`/`avi` into SLEAP, then SLEAP will just reference/point-to that `mp4`/`avi`. Unfortunately, if you imported a dataset as a set of images (i.e. import from DLC), then SLEAP does not currently support outputting an `mp4`/`avi`.

Thanks,
Liezl

**Comment by tctco** [comment]
I see that you have made your datasets publicly available, but they are created for `sleap` specifically and I don't know where I can find the *original videos* that generate these datasets. My algorithm depends on the spatial-temporal info and couldn't infer the identity of animals directly from images :(

**Reply by roomrys** [reply]
Hi @tctco,

Ah the question makes more sense now. The datasets listed on the website are SLEAP Package files denoted by the `.pkg.slp` which - as you noticed - have the frames embedded into the project rather than using a standalone video. There is a [dataset](https://github.com/talmolab/sleap-datasets) (not listed on the website, but found if google "sleap datasets") which contains a video as well.

Thanks,
Liezl

**Reply by tctco** [reply]
Thanks a lot!!!!

BTW, is there an API that I can export the labeled frames to images and files containing all labeled points (just like the DeepLabCut datasets)?

**Reply by roomrys** [reply]
Hi @tctco,

There is not an official API to export labeled frames to images and files containing all labeled points. SLEAP exports a [SLEAP Analysis](https://sleap.ai/notebooks/Analysis_examples.html) file that exports all the labeled/predicted points (preferring user-labeled points). SLEAP also exports a SLEAP Package file for a compact version of the project. 

However, if you need a specific format, you can create a script to export through the command line (in the SLEAP conda environment)
```py
import sleap
labels = sleap.Labels.load_file("path_to_labels/slp_project.slp")
labeled_frames = labels.labeled_frames

# Extract all the images for every labeled frame
for lf in labeled_frames:
     # Get the image for each LabeledFrame
     img = lf.image  # Returns np.ndarray

     # Add some code to save the image here...
     # Your code here

     # Add some code to add the image filename to a csv here...
     # Your code here

# Construct a numpy array from instance points
tracks = labels.numpy()  # Returns shape `(n_frames, n_tracks, n_nodes, 2)`

# Add some code to write the points to a csv here...
# Your code here
```

The SLEAP Classes that will be useful here are [`LabeledFrame`](https://sleap.ai/api/sleap.instance.html#sleap.instance.LabeledFrame) and [`Labels`](https://sleap.ai/api/sleap.io.dataset.html#sleap.io.dataset.Labels). With the template above, you should just need to add your own code. Some third-party libraries ([`pandas.DataFrame.to_csv`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html) might be helpful along with [`PIL.Image.fromarray`](https://pillow.readthedocs.io/en/stable/reference/Image.html?highlight=from_array#PIL.Image.fromarray)).

Let us know if you need any help/clarification.

Thanks,
Liezl

**Reply by tctco** [reply]
Thanks!

**Reply by STAR-811** [reply]
Hi @roomrys! According to the code you offer, the labeled images are exported from the `.slp` file. However, the exported neighboring frames weren't taken at an adjacent time point. So, I can't finish pose tracking through this discontinuous video. The example images are shown below. And I also want to ask @tctco how you solve this problem.
**T=1:**
![image](https://user-images.githubusercontent.com/53425563/185031650-916d4a69-7402-487e-8e40-85d4842315c2.png)
**T=2:**
![image](https://user-images.githubusercontent.com/53425563/185031723-145ef44a-9b54-467b-8966-f7b623b0afc0.png)

**Reply by roomrys** [reply]
Hi @STAR-811,

I converted you issue on this to a discussion (see #917) and will follow up there. In summary, the exported labeled frames are exported in the order they were labeled - I added a bit of code in #917 to help order those labeled frames. Note that even after ordering based on video/frame index, the labeled frames may still not be adjacent as we only export the frames that were labeled in `.pkg.slp` files to help condense the file.

Thanks,
Liezl
